### PR TITLE
Initial f16 support

### PR DESCRIPTION
**Pull request by umar456**
Initial support for fp16:

- [x] JIT
- [x] Random Number Generation
- [x] matmul
- [x] reductions
- [x] reorder/transpose
- [x] printing
- [x] casts
- [x] tests
- ~[ ] convolution~
- ~[ ] anything else~

### CODE REVIEW

**Comment by WilliamTambellini** [comment]
why is there 
"
Initial support for fp16:
CPU
"
in the description ?

**Comment by umar456** [comment]
I was thinking of making a table but thought against it. Fixed the message.

File: CMakeModules/InternalUtils.cmake [hunk 1]
```diff
@@ -104,7 +104,7 @@ macro(arrayfire_set_cmake_default_variables)
   set(CMAKE_PREFIX_PATH "${ArrayFire_BINARY_DIR};${CMAKE_PREFIX_PATH}")
   set(BUILD_SHARED_LIBS ON)
 
-  set(CMAKE_CXX_STANDARD 11)
+  set(CMAKE_CXX_STANDARD 14)
```
Comment (9prady9) [inline CMakeModules/InternalUtils.cmake]: 👍

File: src/backend/cuda/blas.cpp [hunk 2]
```diff
@@ -229,7 +236,7 @@ Array<T> matmul(const Array<T> &lhs, const Array<T> &rhs, af_mat_prop optLhs,
                 getType<T>(), lStrides[1], rhs.get(), getType<T>(), rStrides[1],
                 &beta, out.get(), getType<T>(), out.strides()[1],
                 getType<T>(),  // Compute type
-                CUBLAS_GEMM_DEFAULT_TENSOR_OP));
+                selectGEMMAlgorithm()));
```
Comment (9prady9) [inline src/backend/cuda/blas.cpp]: This change should actually be a separate commit by itself when we merge this PR. Leaving a comment here so that we don't forget to reorganize the commits.

File: src/backend/cuda/CMakeLists.txt [hunk 3]
```diff
@@ -53,7 +53,9 @@ cuda_include_directories(
 set(jit_kernel_headers
     "kernel_headers")
 
-file(GLOB jit_src "kernel/jit.cuh")
+set(jit_src "${CMAKE_CURRENT_SOURCE_DIR}/kernel/jit.cuh;${CUDA_INCLUDE_DIRS}/cuda_fp16.h;${CUDA_INCLUDE_DIRS}/cuda_fp16.hpp")
+
+message(STATUS ${jit_src})
```
Comment (9prady9) [inline src/backend/cuda/CMakeLists.txt]: Debug message

File: src/backend/cuda/kernel/memcopy.hpp [hunk 4]
```diff
@@ -86,11 +87,10 @@ void memcopy(T *out, const dim_t *ostrides, const T *in, const dim_t *idims,
     POST_LAUNCH_CHECK();
 }
 
-////////////////////////////// BEGIN - templated help functions for copy_kernel
-///////////////////////////////////
+///////////// BEGIN - templated help functions for copy_kernel /////////////////
```
Comment (9prady9) [inline src/backend/cuda/kernel/memcopy.hpp]: I thought we are removing these block comments whenever we are editing surrounding code.

File: src/backend/opencl/err_opencl.hpp [hunk 5]
```diff
@@ -22,10 +23,13 @@
...
 template<typename T>
-void verifyDoubleSupport() {
+void verifyTypeSupport() {
     if ((std::is_same<T, double>::value || std::is_same<T, cdouble>::value) &&
         !isDoubleSupported(getActiveDeviceId())) {
         AF_ERROR("Double precision not supported", AF_ERR_NO_DBL);
+    } else if (std::is_same<T, common::half>::value &&
+               !isHalfSupported(getActiveDeviceId())) {
+        AF_ERROR("Half precision not supported", AF_ERR_NO_HALF);
     }
```
Comment (9prady9) [inline src/backend/opencl/err_opencl.hpp]: My comment on #2415 , https://github.com/arrayfire/arrayfire/pull/2415#discussion_r249435163 , is more of concern  especially in OpenCL backend, don't you think so ? We can't assume if double support is there, there is half support.

Hm, is this okay for now because we only ammulate half computation using float for actual computation ?

File: src/backend/common/half.hpp [hunk 6]
```diff
@@ -77,7 +76,7 @@ CONSTEXPR_DH uint16_t float2half(float value) {
     uint32_t bits = 0;  // = *reinterpret_cast<uint32*>(&value);
                         // //violating strict aliasing!
     std::memcpy(&bits, &value, sizeof(float));
-    constexpr uint16_t base_table[512] = {
+    uint16_t base_table[512] = {
```
[the remark this answers is not in the document]
Comment (umar456) [reply src/backend/common/half.hpp]: Hmm. I should make that a static const array. The reason I removed constexpr is because of the memcpy later in the function. The memcpy is not a constexpr operation so the constexpr decorator would have been deceptive here. Also one of the compilers threw an error.

[the remark this answers is not in the document]
Comment (9prady9) [reply src/backend/common/half.hpp]: Got it. Then `static const` should do the trick I guess.

File: src/backend/cuda/blas.cpp (hunk 2 above)
Comment (umar456) [reply src/backend/cuda/blas.cpp]: We aren't using that function before this change. Additionally that function is not compatible with Kepler. Need to find another way to call it.

File: src/backend/cuda/CMakeLists.txt (hunk 3 above)
Comment (umar456) [reply src/backend/cuda/CMakeLists.txt]: Yes. I will need to remove it

File: src/api/c/blas.cpp [hunk 7]
```diff
@@ -157,6 +162,11 @@ af_err af_matmul(af_array *out, const af_array lhs, const af_array rhs,
             case c32: output = matmul<cfloat>(lhs, rhs, optLhs, optRhs); break;
             case f64: output = matmul<double>(lhs, rhs, optLhs, optRhs); break;
             case c64: output = matmul<cdouble>(lhs, rhs, optLhs, optRhs); break;
+#ifndef AF_CPU
+            case f16:
+                output = matmul<common::half>(lhs, rhs, optLhs, optRhs);
+                break;
+#endif
```
[the remark this answers is not in the document]
Comment (umar456) [reply src/api/c/blas.cpp]: You get a type error. The only way to support this is by creating our own gemm implementation.

File: src/backend/cuda/blas.cpp (hunk 2 above)
Comment (9prady9) [reply src/backend/cuda/blas.cpp]: Nevertheless, it is not fp16 related change. Hence, my comment.

File: src/backend/opencl/err_opencl.hpp (hunk 5 above)
Comment (9prady9) [reply src/backend/opencl/err_opencl.hpp]: Nevermind, I missed the type match validation. My bad.

File: include/af/half.h [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+#pragma once
+
+typedef struct {
+    unsigned short data_:16;
+    struct {
+        unsigned short fraction:10;
+        unsigned short exponent:5;
+        unsigned short sign:1;
+    };
+} af_half;
```
Comment (pavanky) [inline include/af/half.h]: Where is this from ?

File: CMakeModules/InternalUtils.cmake [hunk 9]
```diff
@@ -30,7 +30,7 @@ endfunction()
 
 function(arrayfire_get_cuda_cxx_flags cuda_flags)
   if(NOT MSVC)
-    set(flags "-std=c++11 -Xcompiler -fPIC -Xcompiler ${CMAKE_CXX_COMPILE_OPTIONS_VISIBILITY}hidden")
+    set(flags "-std=c++14 -Xcompiler -fPIC -Xcompiler ${CMAKE_CXX_COMPILE_OPTIONS_VISIBILITY}hidden")
```
Comment (WilliamTambellini) [inline CMakeModules/InternalUtils.cmake]: are you such a change/move should nt be split into a sub separate PullRequest "upgrade to std c++14" ?

File: include/af/device.h [hunk 10]
```diff
@@ -75,10 +75,21 @@ namespace af
...
 
+    /// \brief Queries the current device for half precision floating point
+    ///        support
+    ///
+    /// \param[in] device the ID of the device to query
+    ///
+    /// \returns true if the \p device supports half precision operations.
+    ///          false otherwise
+    /// \ingroup device_func_half
+    AFAPI bool isHalfAvailable(const int device);
```
Comment (WilliamTambellini) [inline include/af/device.h]: Could you add more comments in order to give some examples of which device will return true which false ?

File: src/backend/cuda/platform.cpp [hunk 11]
```diff
@@ -203,6 +203,12 @@ bool isDoubleSupported(int device) {
     return true;
 }
 
+bool isHalfSupported(int device) {
+    auto prop = getDeviceProp(device);
+    float compute = prop.major * 1000 + prop.minor * 10;
+    return compute >= 5030;
```
Comment (WilliamTambellini) [inline src/backend/cuda/platform.cpp]: hum does it mean that any device setup with CudaComp 5.3+ will really support half ?

File: src/backend/cuda/blas.cpp [hunk 12]
```diff
@@ -179,21 +287,25 @@ void gemm(Array<T> &out, af_mat_prop optLhs, af_mat_prop optRhs,
     if (oDims.ndims() <= 2) {
         if (rDims[bColDim] == 1) {
             dim_t incr = (optRhs == AF_MAT_NONE) ? rStrides[0] : rStrides[1];
+            if (is_same<T, half>::value) {
+                AF_ERROR(
+                    "GEMV does not support half, Please create an issue on "
```
Comment (WilliamTambellini) [inline src/backend/cuda/blas.cpp]: is it impossible to gemv f16 arrays or just to be implemented ?

File: src/backend/cuda/CMakeLists.txt [hunk 13]
```diff
@@ -450,7 +456,7 @@ cuda_add_library(afcuda
   )
 
 arrayfire_set_default_cxx_flags(afcuda)
-target_compile_definitions(afcuda PRIVATE AF_CUDA)
+target_compile_definitions(afcuda PRIVATE AF_CUDA CUDA_NO_HALF)
```
Comment (WilliamTambellini) [inline src/backend/cuda/CMakeLists.txt]: Could you please add a comment what CUDA_NO_HALF does ?

File: CMakeModules/InternalUtils.cmake (hunk 9 above)
Comment (umar456) [reply CMakeModules/InternalUtils.cmake]: Yeah, Thats a good idea. I can do that.

File: src/backend/cuda/platform.cpp (hunk 11 above)
Comment (umar456) [reply src/backend/cuda/platform.cpp]: Yes, only those devices have support for half operations. I can try to incorporate older devices but that will take more time.

File: src/backend/cuda/blas.cpp (hunk 12 above)
Comment (umar456) [reply src/backend/cuda/blas.cpp]: It will have to be implemented internally. Its not something that is part of cuBLAS

File: src/backend/cpu/cast.hpp [hunk 14]
```diff
@@ -26,6 +26,112 @@ struct UnOp<To, Ti, af_cast_t> {
...
+        for (int i = 0; i < lim; i++) {
+            float val = in[i];
+            out[i]    = To(val);
+        }
+    }
+
+    void eval(jit::array<To> &out, const jit::array<float> &in, int lim) {
+        for (int i = 0; i < lim; i++) { out[i] = To(in[i]); }
+    }
+};
```
Comment (9prady9) [inline src/backend/cpu/cast.hpp]: The specialization for `std::complex<float>` output type for `struct UnOp<To, common::half, af_cast_t>` looks exactly same in code to the general implementation. Doesn't look like we need it from `half` to `To`.

File: src/backend/cpu/cast.hpp [hunk 15]
```diff
@@ -26,6 +26,112 @@ struct UnOp<To, Ti, af_cast_t> {
...
```
Comment (9prady9) [inline src/backend/cpu/cast.hpp]: This specialization also looks identical to the general implementation of `struct UnOp<To, common::half, af_cast_t>`.

File: src/backend/cpu/kernel/transpose.hpp [hunk 16]
```diff
@@ -12,6 +12,8 @@
 #include <err_cpu.hpp>
 #include <utility.hpp>
 
+using af::dim4;
```
Comment (9prady9) [inline src/backend/cpu/kernel/transpose.hpp]: I thought we stopped adding `using` statements in headers ?

File: src/backend/cuda/blas.cpp [hunk 17]
```diff
@@ -150,14 +156,116 @@ BLAS_FUNC(dot, cdouble, false, Z, u)
...
+    } else {
+        using Nt = typename common::kernel_type<T>::native;
+        return gemm_func<Nt>()(blasHandle(), lOpts, rOpts, M, N, K, (Nt *)alpha,
+                               (Nt *)lhs.get(), lStride, (Nt *)rhs.get(),
+                               rStride, (Nt *)beta, (Nt *)out.get(), oleading);
+    }
+}
+
+template<typename T>
+cublasStatus_t gemmBatched_dispatch(BlasHandle handle,
```
Comment (9prady9) [inline src/backend/cuda/blas.cpp]: I don't like this mixed naming convention. `gemmBatched_dispatch` should be `gemmBatchedDispatch` and the other one should be `gemmDispatch`.

File: src/backend/opencl/kernel/transpose.cl [hunk 18]
```diff
@@ -15,6 +15,12 @@ T doOp(T in) {
 #define doOp(in) in
 #endif
 
+#pragma OPENCL EXTENSION all : enable
+#ifdef cl_khr_fp16
+#else
+#define half short
+#endif
```
Comment (9prady9) [inline src/backend/opencl/kernel/transpose.cl]: Interesting, why is it that we need these `#if`s only in `jit.cl` and `transpose.cl` ?

File: test/array.cpp [hunk 19]
```diff
@@ -335,6 +335,16 @@ TYPED_TEST(Array, TypeAttributes) {
...
             break;
+        case f16:
+            EXPECT_TRUE(one.isfloating());
+            EXPECT_FALSE(one.isdouble());
+            EXPECT_FALSE(one.issingle());
+            EXPECT_TRUE(one.isrealfloating());
+            EXPECT_FALSE(one.isinteger());
+            EXPECT_TRUE(one.isreal());
+            EXPECT_FALSE(one.iscomplex());
+            EXPECT_FALSE(one.isbool());
```
Comment (9prady9) [inline test/array.cpp]: don't we have `isHalf` now ?

File: test/join.cpp [hunk 20]
```diff
@@ -44,7 +44,7 @@ class Join : public ::testing::Test {
 
 // create a list of types to be tested
 typedef ::testing::Types<float, double, cfloat, cdouble, int, unsigned int,
-                         intl, uintl, char, unsigned char, short, ushort>
+                         intl, uintl, char, unsigned char, short, ushort, af_half>
```
Comment (9prady9) [inline test/join.cpp]: we should be C++ typedef `af::half` to be consistent with `cfloat` and `cboudle`. This applies to all the tests where `half` is enabled.

File: test/testHelpers.hpp [hunk 21]
```diff
@@ -537,6 +576,10 @@ const af::cfloat &operator+(const af::cfloat &val) { return val; }
 
 const af::cdouble &operator+(const af::cdouble &val) { return val; }
 
+const af_half& operator+(const af_half& val) {
```
Comment (9prady9) [inline test/testHelpers.hpp]: `af::half` to be consistent with other `af::*` types.

File: test/testHelpers.hpp [hunk 22]
```diff
@@ -34,6 +35,16 @@
 #include <mmio.h>
 #endif
 
+bool operator==(const af_half &lhs, const af_half &rhs) {
+    return lhs.data_ == rhs.data_;
+}
+
+std::ostream &operator<<(std::ostream &os, const af_half &val) {
```
Comment (9prady9) [inline test/testHelpers.hpp]: Do we use `half_float::half` from `extern/half/include/half.hpp` only in tests ?

File: src/backend/opencl/kernel/transpose.cl (hunk 18 above)
Comment (umar456) [reply src/backend/opencl/kernel/transpose.cl]: We will probably need it in other places but those functions haven't been implemented yet.

File: test/testHelpers.hpp (hunk 22 above)
Comment (umar456) [reply test/testHelpers.hpp]: Yes. Our half type is not fully functional. It would require us to reimplement the entire half library inorder to correctly implement it. I don't think we should directly include other libraries in our API. perhaps we can discuss it in a separate issue.

File: test/join.cpp (hunk 20 above)
Comment (umar456) [reply test/join.cpp]: I am worried we will have conflicts with the half data type. I want to discourage the use of this type by users because of its inefficiency. I am not sure if this is the best way to handle this sort of thing.

File: include/af/device.h (hunk 10 above)
Comment (WilliamTambellini) [reply include/af/device.h]: On my side I m more interested to know if the device can handle half natively with better perf than f32 (most of the time meaning Volta/Turing via TensorCores and apparently P100). 
Is nt that method going to return true on old GPUs (GTX10X0) where half is perhaps "supported" but perf is disastrous ?

Comment (umar456) [reply include/af/device.h]: The half support for the 10x0 cards are supported through floats.

Even on the Volta cards we need to improve the way we handle things to improve the compute performance. In order to fully utilize the GPU's half compute performance we need to use half2. This will require a major shift on some of the internal kernels so I pushed that change off to later. We can talk about what functions are important to you and improve those later down the line. All the library calls for half(i.e. matmul) will be using half2 so you will get full compute performance there. If there are kernels that are bandwidth bound moving to half2 will serve no purpose because it will still be a bandwidth bound kernel.

File: test/array.cpp [hunk 23]
```diff
@@ -335,6 +335,17 @@ TYPED_TEST(Array, TypeAttributes) {
...
+            EXPECT_FALSE(one.ishalf());
```
Comment (syurkevi) [inline test/array.cpp]: EXPECT_TRUE?
